subtests/docker_cli/systemd/systemd.py

```diff
@@ -1,86 +1,155 @@
-import time
-import socket
 from os.path import exists
+from os.path import join
+from os import rename
 from os import unlink
+from shutil import copyfile
 from autotest.client import utils
-from autotest.client.shared.utils import is_port_free
-from dockertest.subtest import Subtest
-from dockertest.dockercmd import DockerCmd
+from dockertest.subtest import SubSubtest
+from dockertest.subtest import SubSubtestCaller
 from dockertest.images import DockerImages
-from dockertest.output import mustpass
 from dockertest.xceptions import DockerTestError
 
 
-class systemd(Subtest):
+class systemd(SubSubtestCaller):
+    pass
+
+
+class systemd_base(SubSubtest):
 
     def initialize(self):
-        super(systemd, self).initialize()
-        unit_file_srcpath = '{}{}'.format(self.bindir, '/p4321.service')
-        shutil.copyfile(unit_file_srcpath, self.config['sysd_unitf_dest'])
-        dkrcmd = DockerCmd(self, 'build', ['--force-rm -t',
-                                           self.config['name'], self.bindir])
-        mustpass(dkrcmd.execute())
+        super(systemd_base, self).initialize()
+        self.sub_stuff['bindir'] = self.parent_subtest.bindir
+        self.sub_stuff['tmpdir'] = self.tmpdir
+        unitfile = self.config['sysd_unit_file']
+        self.sub_stuff['unitf_dst'] = join('/etc/systemd/system', unitfile)
+        self.sub_stuff['tmpdirf'] = join(self.sub_stuff['tmpdir'], unitfile)
+        self.sub_stuff['dkrfl'] = join(self.sub_stuff['bindir'], 'Dockerfile')
+        copyfile(join(self.sub_stuff['bindir'], unitfile),
+                 self.sub_stuff['tmpdirf'])
+        self.sub_stuff['vars_dict'] = {'name': self.config['image_name']}
 
     def run_once(self):
-        host = 'localhost'
-        port = 4321
         self.sysd_action('daemon-reload')
-        self.sysd_action('start', self.config['sysd_unit_file'])
-        utils.wait_for(lambda: not is_port_free(port, host), 10)
-        time_from_socket = self.read_socket(host, port)
-        self.failif(not time_from_socket.isdigit(),
-                    "Data received from container is non-numeric: '%s'"
-                    % time_from_socket)
+        for act in 'start', 'status':
+            self.sysd_action(act, self.config['sysd_unit_file'])
+
+    def postprocess(self):
+        super(systemd_base, self).postprocess()
+        img_lst = ''
+        for img in DockerImages(self).list_imgs_full_name():
+            img_lst = '{} {}'.format(img_lst, img.split('/', -1)[-1])
+        if self.config['image_name'] not in img_lst:
+            raise DockerTestError("Image %s is not present"
+                                  % self.config['image_name'])
 
     def cleanup(self):
-        super(systemd, self).cleanup()
-        filepath = self.config['sysd_unitf_dest']
+        super(systemd_base, self).cleanup()
+        filepath = self.sub_stuff['unitf_dst']
         if exists(filepath):
             self.sysd_action('stop', self.config['sysd_unit_file'])
```

Comment:
If ``super(...)`` or ``sysd_action`` fails for any reason.  The unit file won't be deleted.  I think you need a ``try:...finally:...`` here to be extra sure the file gets removed and daemon is reloaded.